subsys/bluetooth/audio/vocs.c

```diff
@@ -0,0 +1,482 @@
+			     void *buf, uint16_t len, uint16_t offset)
+{
+	struct vocs_server *inst = attr->user_data;
+
+	BT_DBG("0x%02x", inst->location);
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &inst->location,
+				 sizeof(inst->location));
+}
+#endif /* CONFIG_BT_VOCS */
+
+static ssize_t write_vocs_control(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				  const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+	const struct vocs_control_t *cp = buf;
+	bool notify = false;
+
+	if (!len || !buf) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	/* Check opcode before length */
+	if (!VALID_VOCS_OPCODE(cp->opcode)) {
+		BT_DBG("Invalid opcode %u", cp->opcode);
+		return BT_GATT_ERR(BT_VOCS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (offset) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len != sizeof(struct vocs_control_t)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	BT_DBG("Opcode %u, counter %u", cp->opcode, cp->counter);
+
+
+	if (cp->counter != inst->state.change_counter) {
+		return BT_GATT_ERR(BT_VOCS_ERR_INVALID_COUNTER);
+	}
+
+	switch (cp->opcode) {
+	case VOCS_OPCODE_SET_OFFSET:
+		BT_DBG("Set offset %d", cp->offset);
+		if (cp->offset > BT_VOCS_MAX_OFFSET || cp->offset < BT_VOCS_MIN_OFFSET) {
+			return BT_GATT_ERR(BT_VOCS_ERR_OUT_OF_RANGE);
+		}
+
+		if (inst->state.offset != sys_le16_to_cpu(cp->offset)) {
+			inst->state.offset = sys_le16_to_cpu(cp->offset);
+			notify = true;
+		}
+		break;
+	default:
+		return BT_GATT_ERR(BT_VOCS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (notify) {
+		inst->state.change_counter++;
+		BT_DBG("New state: offset %d, counter %u",
+		       inst->state.offset, inst->state.change_counter);
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_STATE, inst->service_p->attrs,
+				    &inst->state, sizeof(inst->state));
+
+		if (inst->cb && inst->cb->state) {
+			inst->cb->state(NULL, (struct bt_vocs *)inst, 0, inst->state.offset);
+		}
+
+	}
+
+	return len;
+}
+
+#if defined(CONFIG_BT_VOCS)
+static void output_desc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+#endif /* CONFIG_BT_VOCS */
+
+static ssize_t write_output_desc(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+
+	if (len >= sizeof(inst->output_desc)) {
+		BT_DBG("Output desc was clipped from length %u to %zu",
+		       len, sizeof(inst->output_desc) - 1);
+		/* We just clip the string value if it's too long */
+		len = (uint16_t)sizeof(inst->output_desc) - 1;
+	}
+
+	if (len != strlen(inst->output_desc) || memcmp(buf, inst->output_desc, len)) {
+		memcpy(inst->output_desc, buf, len);
+		inst->output_desc[len] = '\0';
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_DESCRIPTION, inst->service_p->attrs,
```

Comment:
Check return value?

Reply:
What do you suggest we do in case of error? Logging with BT_DBG or BT_WARN? 

There's not really anything besides logging we can do as the server.

Reply:
I was mostly asking :) I suggest adding an explicit ignore of the return value.
`(void)bt_gatt_notify_uuid(...)`